common/src/main/java/org/apache/rocketmq/common/LifecycleAwareServiceThread.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.rocketmq.common;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class LifecycleAwareServiceThread extends ServiceThread {
+
+    private final AtomicBoolean started = new AtomicBoolean(false);
+
+    @Override
+    public void run() {
+        started.set(true);
+        synchronized (started) {
+            started.notifyAll();
+        }
+
+        run0();
+    }
+
+    public abstract void run0();
+
+    public void awaitStarted(long timeout) throws InterruptedException {
+        synchronized (started) {
+            if (started.get()) {
+                return;
+            }
+            started.wait(timeout);
```

Comment:
It seems that the wait here could not ensure that the service is started successfully.

On the other hand, how could we avoid spurious wakeup?

Reply:
OK